jobbergate-core/jobbergate_core/tools/sbatch.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import annotations
+
+import json
+import re
+import shlex
+import subprocess
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Iterable
+
+from buzz import check_expressions
+from loguru import logger
+
+
+def inject_sbatch_params(job_script_data_as_string: str, sbatch_params: list[str], header: str | None = None) -> str:
+    """
+    Inject sbatch params into job script.
+
+    Given the job script as job_script_data_as_string, inject the sbatch params in the correct location.
+    """
+    logger.debug("Preparing to inject sbatch params into job script")
+
+    if not sbatch_params:
+        logger.warning("Sbatch param list is empty")
+        return job_script_data_as_string
+
+    # Find the first non-blank, non-comment line
+    match = re.search(r"^[^#\n]", job_script_data_as_string, re.MULTILINE)
+    if match:
+        insert_index = match.start()
+    else:
+        # If no such line is found, append at the end
+        insert_index = len(job_script_data_as_string)
+
+    inner_string = f"# {header}\n" if header else ""
+    for parameter in sbatch_params:
+        inner_string += f"#SBATCH {parameter}\n"
+    else:
+        inner_string += "\n"
+
+    new_job_script_data_as_string = (
+        job_script_data_as_string[:insert_index] + inner_string + job_script_data_as_string[insert_index:]
+    )
+
+    logger.debug("Done injecting sbatch params into job script")
+    return new_job_script_data_as_string
+
+
+@dataclass(frozen=True)
+class SbatchHandler:
+    """Submits sbatch jobs to the cluster."""
+
+    username: str
+    sbatch_path: Path
+    scontrol_path: Path
+    submission_directory: Path = field(default_factory=Path)
+
+    def __post_init__(self):
+        with check_expressions("Check paths", raise_exc_class=ValueError) as check:
+            check(self.sbatch_path.is_absolute(), "sbatch_path is not an absolute path")
+            check(self.sbatch_path.exists(), "sbatch_path does not exist")
+            check(self.scontrol_path.is_absolute(), "scontrol_path is not an absolute path")
+            check(self.scontrol_path.exists(), "scontrol_path does not exist")
+
+    def run(self, job_script_path: Path) -> int:
+        """Runs sbatch as the user."""
+        command = (
+            self.sbatch_path.as_posix(),
+            job_script_path.as_posix(),
+            "--parsable",
+        )
+
+        completed_process = self._run_command_as_user(
+            command, cwd=self.submission_directory, capture_output=True, text=True
+        )
+
+        try:
+            return int(completed_process.stdout.split(",")[0])
```

Comment:
It would be good to indicate here a little more what is being parsed. Also, it would be good to indicate that the return value is a slurm job id.

Reply:
Sure, I'll clarify it.

Reply:
solved in [19be649](https://github.com/omnivector-solutions/jobbergate/pull/498/commits/19be6495094d96b2ede2b7900dede98fc500000b)